### ISSUE TITLE
Add Maintenance UI View / Window

### ISSUE THREAD

**Issue by noah-paige** [data-dot-all/dataall#998]
**Is your idea related to a problem? Please describe.**
Need a way to describe what new features had recently been introduced to data.all on latest releases and if there should be any warnings, deprecation notices, etc.

**Describe the solution you'd like**
To communicate new features or maintenance actions - add a maintenance UI view or window where we can post messages and info about the current status of the platform etc..

*P.S. Don't attach files. Please, prefer add code snippets directly in the message body.*

**Comment by zsaltys** [comment]
My proposal to implement this:

1) add a settings page which is only visible to data all administrators
2) on this settings page add a button to put data.all into maintenance mode on/off.. store the state in RDS.
3) when maintenace mode is on all users except da admins see maintenance window
4) during maintenance mode all graphql calls are blocked except those needed to make maintenance mode work
5) no new cron jobs / periodic jobs should start while maintenance mode is on.. no stacks should be updated.
6) da admins only see the settings page while maintenance mode is on.. cannot navigate to anything else

**Comment by TejasRGitHub** [comment]
Hi @zsaltys. Thanks for the insight. 

**WIP**
From the description by Noah on this issue .We also need a place to show what's in the new release 

Where do we get this information about new features ? 
One place could be from the release notes of the UI Or make another readme/ text file which will contain information about the new release 

Where should this UI be ? 

Do we wan to keep a time limit to how long we want to show this UI ? 

**Maintenance Window** 

The maintenance window UI which will be only visible to the data.all admin group could be added in the Admin settings page as another tab.
<img width="1479" alt="image" src="https://github.com/data-dot-all/dataall/assets/71188245/8fcda7a9-ff79-47fd-a8f2-889430885eac">

The UI will show a button to create a maintenance window and a table showing past maintenance windows. 
When clicked the button will show a pop-up to confirm the actions and then will start the maintenance window. The user will be shown if its okay to start the deployment ( For e.g. when clicked the maintenance window will show IN-PROGRESS and then once it is confirmed that no more pending graphQL calls, stack updates remain then the UI will  change to READY-FOR-DEPLOYMENT. Once the deployment is done the data.all admin can then remove the maintenance window ) 

A `maintenance` table will have to be created with attributes like 
User - Who started the window
Start time
End time 
Status of maintenance window 

List of tasks that need to be shut off for maintenance window 
1. GraphQL calls will be blocked when the maintenance window is ON -> Sent unauthorized ? if maintenance window is present  Or should the API Gateway be turned off / paused for a while
2. Pause all the scheduled ECS tasks -> ECS stacks Updater, ECS table syncer, ECS Subscriptions , ECS catalog index syncer, ECS policy updater 
3. Check all the currently running Stack updates. and wait for them to finish  . Tasks include dataset stack update, environment stack update, ( any other ?  ) 

UI shown to the user when the maintenance window is scheduled 
When the maintenance window is ON, the user will just be shown a blank UI with a custom text about maintenance window. This custom text could be picked up from the config.json. ( Add this UI in the AuthGuard which is triggered for each Navigation a user does ? )  


Question to answer - 
Q1 . How to effectively shutoff the GraphQL API calls. Do it programatically inside the graphQL lambda or switch of the API gateway ? 

Q2. How to pause the ECS scheduled tasks ? 

What about users who are already logged in ? Once the maintenance window is on the graphQL calls will fail. We may want the user to logout and login back in to see the maintenance UI 

How to make sure all the ECS and cron scheduled jobs are shut off ?

**Comment by anushka-singh** [comment]
One other thing to put on the display would be a timer or information on when we expect the system to be back up. Not having that information could cause confusion and anxiousness to users.

**Comment by anushka-singh** [comment]
I have some questions: 
1. How will the system handle long-running tasks that are in progress when the maintenance window begins? Will they be allowed to complete, or will they be forcibly terminated?
2. What monitoring and alerting mechanisms will be in place to notify administrators of any issues or anomalies during maintenance windows? How will the system handle unexpected errors or failures?
3. What procedures will be in place for rolling back changes in the event of unexpected issues or failures during maintenance windows? 
4. We should also have a way to notify users in advance of an upcoming maintenance window - either via email or on the notifications UI in data all or both.

**Comment by rbernotas** [comment]
I'm not sure we need to store user/times in RDS in a new table.  If we need to store that for any reason it could conceivably be stored in the activity table as a single additional record.  Alongside that, I wouldn't recommend setting specific start/end times of the maintenance window within data.all.  data.all should show that a maintenance window is happening, perhaps having some custom configurable text shown (that may redirect users to some other internal communications like Slack/etc which is giving more details about any maintenance).  

We may need to consider that admins will still need access to the UI/GraphQL during the maintenance window, for purposes of validation and/or debugging.  Upgrades/deploys also may not be the only reason we need maintenance windows where we need to block users from using the data.all instance.  Certainly regular users should be blocked, but ideally data.all admins can still do whatever they want with data.all.

**Comment by TejasRGitHub** [comment]
> One other thing to put on the display would be a timer or information on when we expect the system to be back up. Not having that information could cause confusion and anxiousness to users.

As the deployments do have variable time, having a timer which can be set on the maintenance UI will be tricky and in some places the data.all admin will have to extend it. Instead , a custom text - containing communication about the maintenance window and also who to contact , etc - can be shown ( which can be set in the config.json file  ) on the UI when the maintenance is taking place.

**Comment by TejasRGitHub** [comment]
> I have some questions:
> 
> 1. How will the system handle long-running tasks that are in progress when the maintenance window begins? Will they be allowed to complete, or will they be forcibly terminated?
> 2. What monitoring and alerting mechanisms will be in place to notify administrators of any issues or anomalies during maintenance windows? How will the system handle unexpected errors or failures?
> 3. What procedures will be in place for rolling back changes in the event of unexpected issues or failures during maintenance windows?
> 4. We should also have a way to notify users in advance of an upcoming maintenance window - either via email or on the notifications UI in data all or both.

For 
Q1. The maintenance window status will be in READY-FOR-DEPLOYMENT once the long running and any pending stacks which were running are completed. When the maintenance window is triggered on the UI, the subsequent graphQL calls will be blocked but all the running stack updates will be monitored till they are completed. 

Q2. The purpose of the maintenance window would be just to create a safe window to deploy. In case of deployment failure, admins, devs will have to monitor and fix the deployment error. Although, it would be great to have some automation in case of failure it would be a larger development and out of scope of this task. 

Q3. For any failures and roll backs, the deployment team will have to intervene and fix those issue. 

Q4. I like this idea of notification to the users about the maintenance window. I will take this into consideration will making the design. Thanks!

**Comment by noah-paige** [comment]
From internal discussion with @TejasRGitHub:

1. During  maintenance window - only block mutations instead of blocking all calls to graphQL (for non DA Admins)
  - No need for a new custom page to route users to during maintenance window (adds complexity to handle both current users logged in to dataall UI and new users logging in routed to a separate page)
  - Can keep UI the same and block ALL mutations (unless DA Admin)
  - Can have notification banner that just tells user that maintenance window is active

2. Check if we need to poll on ECS and if its worth the developer effort.
  - I would encourage the design of this PR to not include polling of ECS tasks once a Maintenance Window is enabled
  - Even if the Maintenance Window is in a status of `Pending` what is preventing the DA Admins from upgrading and starting new CodePipeline Execution
  - I think Maintenance Window enabled should just block all NEW mutation API calls and prevent any NEW ECS scheduled tasks (responsibility on admin team to upgrade appropriately)

3. Admin can add custom email notifications on a button present on maintenance window UI
  - Similar to how maintenance window cna be enabled/disabled from admin settings, lets have a way for Admins to broadcast a notification (via UI button or similar) letting all DA Users know about maintenance window
  - Can leverage existing Notifications module code to send notification in both data.all UI and via email using SES (if configured)

4. Link to new features instead of showing up a modal pop-up
  - For new features display can have a button in the header of data.all UI (as shown in image above) that links to the OS Release notes for the latest version (or links to some custom release notes via configuration file?)
  - No need to dynamically read release notes from somewhere or pull in a markdown file - likely easiest to have a link here and will serve same purpose 

5. [Nice to have]  - have a way to hide the link / modal popup for new features UI
  - Look into frontend design to broadcast pop up notification to notify user that an upgrade just happened and to refer to new features for more information
  - Once user acknowledges the pop-up it never shows again

**Comment by dlpzx** [comment]
@TejasRGitHub just to add on the topics above. I agree with the summary from @noah-paige: block mutations not queries, block new tasks/APIs only, DAAdmins can do all tasks. I would add:
- Since we are blocking a lot of actions, maybe we should make it very evident to users that the maintenance window is open. In my opinion, the notification banner should not be a temporal banner like the one we usually get, but a permanent window or a significant UI change (e.g. change some of the theme colors)
- I would like to be able the current data.all version at any moment in the UI, the new features and also THE FUTURE DEPRECATION of features. I think we are not showing this with the current design right?

**Comment by TejasRGitHub** [comment]
> Hi @TejasRGitHub I am reviewing your design, here are my comments:
> 
> * in Phase 1, When in NO-ACCESS mode, all OpenSearch calls are blocked ----> Why are they blocked for the admins?
> * Regarding the decision on the ECS Running Tasks status options ----> I think I am more inclined towards option 2. As far as we provide good error messages when we try to change the maintenance mode I think we are good. At the end, the users of this part of the feature are data.all admins, we assume they have a level of data.all understanding :)
> * Question regarding the schema changes ----> Is the table going to contain a single row with the status and mode of the platform or are we going to create an item for each time we update the status (like a history of maintenance changes)?
> * I like the phases approach

For 1 , Corrected the statement. All calls on opensearch will be blocked except for data.all admins. 
For Question regarding schema, the schema will only contain one row and this will be inserted during the alembic migration. During each maintenance mode, the status of that one row will be updated to reflect the maintenance status at that time. For the history of who did what during maintenance mode , the activity table could be used.

**Comment by anushka-singh** [comment]
@TejasRGitHub 
- Here:
`GraphQL Calls -
These calls will be blocked on mutation requests and others will be allowed. `
Can you give an example of what "others" calls are? Which graphql calls will be allowed for normal users?

- Instead of error message: `'Unauthorized : data.all is in maintenance mode and thus your actions are blocked for now'`  maybe we use: `"Access Restricted: data.all is currently undergoing maintenance, and your actions are temporarily blocked."`

- `Cons - User will have to come to the data.all UI and trigger this polling. If user is not on the UI they will miss polling even when the ECS tasks have completed and maintenance window is in READY-FOR-DEPLOYMENT mode. `
For this: I would call this out as a risk and look for a solution for this. It can often happen that as an admin I started maintenance window and got busy with some other work. I should still have a way to know when the system is READY-FOR-DEPLOYMENT.

**Comment by TejasRGitHub** [comment]
> @TejasRGitHub
> 
> * Here:
>   `GraphQL Calls - These calls will be blocked on mutation requests and others will be allowed. `
>   Can you give an example of what "others" calls are? Which graphql calls will be allowed for normal users?

By Other calls I mean the query calls . I will correct the wording in the proposal. For normal users in READ-ONLY mode, query calls will be allowed and mutation calls will be blocked. In NO-ACCESS Mode, both query as well as mutation calls will be blocked 
> * Instead of error message: `'Unauthorized : data.all is in maintenance mode and thus your actions are blocked for now'`  maybe we use: `"Access Restricted: data.all is currently undergoing maintenance, and your actions are temporarily blocked."`

Thanks for the suggestion. Will use this one instead.
> * `Cons - User will have to come to the data.all UI and trigger this polling. If user is not on the UI they will miss polling even when the ECS tasks have completed and maintenance window is in READY-FOR-DEPLOYMENT mode. `
>   For this: I would call this out as a risk and look for a solution for this. It can often happen that as an admin I started maintenance window and got busy with some other work. I should still have a way to know when the system is READY-FOR-DEPLOYMENT.

I completely understand this problem. In order to address it, I will be providing a note on the maintenance UI which will explicitly tell data.all admin user to check back on the UI for getting updates on maintenance window status. In the future phases, while working on scheduled maintenance window and email notifications this issue can be handled in a better way.

**Comment by dlpzx** [comment]
Close as completed
